packages/loopring_v3/contracts/impl/relayer/Relayer.sol

```diff
@@ -0,0 +1,50 @@
+/*
+
+  Copyright 2017 Loopring Project Ltd (Loopring Foundation).
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+
+import "./RelayerData.sol";
+
+import "../../lib/Claimable.sol";
+import "../../lib/ReentrancyGuard.sol";
+
+
+/// @title An Implementation of IRelayer.
+/// @author Brecht Devos - <brecht@loopring.org>
+contract Relayer is Claimable, ReentrancyGuard, RelayerData
```

Comment:
Relayer => Agenct/Delegate

Reply:
> My concern is that now the definition of conditions is basically the tuple of from, to, token, amount. It serves fast withdrawal well enough, but may not be as generalized as it can be. I was thinking that conditions can be anything programmable using solidity, for example, if address X has a balance bigger than 10000 Ether, then transfer 0.1 Ether from my offchain account to accountId 123. -- It doesn't have a matching onchain transfer the order direction.

Why would it need a matching onchain transfer? What you described can be done like this:

```
function executeConditionalTransfer(
  address targetAddress,
  address to
  )
  external
{
  require(targetAddress.balance >= 10000 eth, "CONDITION_FAILED");

  // Approve the conditional transfer
  address token = 0x0; // ETH
  IExchangeV3(exchangeAddress).approveConditionalTransfer(
    msg.sender,
    to,
    token,
    1 eth,
    token,
    0,
    uint32(now)
  );
}
```

The tuple of `from`, `to`, `token`, `amount` is of the conditional transfer itself, which is completely independent of the logic done on-chain. It can even be a simple fee payment for the gas cost to caller in exchange for doing the  smart contract logic for the user. It doesn't have to match anything onchain, the module can do any logic it wants, it can even allow multiple conditional transfers if needed.

Reply:
> Relayer => Agenct/Delegate

I quite like Agent. :)

Reply:
Some other quick examples of modules:

Using onchain liquidity for simple trades without order matching:
```
function executeOnchainLiquidityTrade(
  address tokenSellerAddress,
  address accountOwner,
  address cookieJarOwner,
  address tokenS,
  address tokenB,
  uint    fAmount,
  uint32  salt,
  bytes   calldata signature
  )
  external
{
  // Check the signature
  bytes32 hash = keccak256(
    abi.encodePacked(
      tokenSellerAddress,
      accountOwner,
      cookieJarOwner,
      tokenS,
      tokenB,
      uint(fAmount).decodeFloat(24),
      salt
    )
  );
  require(hash.toEthSignedMessageHash().recover(signature) == accountOwner, "UNAUTHORIZED");

  // Transfer the tokens here from the cookie jar contract
  ICookieJarContract(cookieJarContractAddress).transfer(
    address(this),
    tokenS,
    uint(fAmount).decodeFloat(24)
  );
  // Alternatively, we could also transfer tokens directly from the user's own wallet like this if possible:
  // onchainTransferFrom(accountOwner, address(this), tokenS, uint(fAmount).decodeFloat(24));

  // Expects the tokens that are sold from 'this' and are also received on 'this' address
  // Approvals may need to be set etc..
  uint amountB = ITokenSeller(tokenSellerAddress).sellToken(
    tokenS,
    tokenB, 
    uint(fAmount).decodeFloat(24)
  );

  // Deposit the tokens from 'this' directly to the user's account
  exchange.depositTo(accountOwner, tokenB, amountB);

  // Approve the conditional transfer
  // (A fee can be charged in any token as well)
  IExchangeV3(exchangeAddress).approveConditionalTransfer(
    accountOwner,
    cookieJarOwner,
    token,
    fAmount,
    token,
    0,
    salt
  );
}
```

Using the Relayer's authorization capabilities to do a user's deposit:

```
function executeDeposit(
  address accountOwner,
  address token,
  uint    fAmount,
  uint32  salt,
  bytes   calldata signature
  )
  external
{
  // Check the signature
  bytes32 hash = keccak256(
    abi.encodePacked(
      accountOwner,
      token,
      uint(fAmount).decodeFloat(24),
      salt
    )
  );
  require(hash.toEthSignedMessageHash().recover(signature) == accountOwner, "UNAUTHORIZED");

  // Deposit the tokens from the user's wallet
  exchange.deposit(accountOwner, token, amount);

  // Charge a fee here for the gas cost with a conditional transfer from 
  // the user's offchain balances or do an onchain transfer from the user's wallet
  // ...
}
```